analysis/ps.R

```diff
@@ -0,0 +1,141 @@
+####
+## This script does the following:
+# 1. Import processed data
+# 2. Create a propensity score for treatment
+# 3. Extract SMDs before and after IPW based on PS
+# 4. Density plot; untrimmed and trimmed
+# 5. Save all output
+####
+
+# Import libraries and functions ------------------------------------------
+library(arrow)
+library(here)
+library(dplyr)
+library(rms) # splines
+library(cobalt) # SMD density
+library(ggplot2)
+
+# Create directories for output -------------------------------------------
+fs::dir_create(here::here("output", "ps"))
+
+# Import the data ---------------------------------------------------------
+df <- read_feather(here("output", "data", "data_processed.arrow"))
+
+# Add splines -------------------------------------------------------------
+# Compute knot locations based on percentiles, according to study protocol
+age_knots <- quantile(df$cov_num_age, probs = c(0.10, 0.50, 0.90))
+# print(age_knots)
+
+# PS model ----------------------------------------------------------------
+# Set format and reference to ensure proper logistic regression modeling
+df$exp_bin_treat <- factor(df$exp_bin_treat, 
+                                       levels = c(0, 1), # Reference first
+                                       labels = c("nothing", "metformin"))
+
+# Define the covariates included in the PS model
+covariate_names <- names(df) %>%
+  grep("^cov_", ., value = TRUE) %>% 
+  # exclude those not needed in the model:
+  ## cov_bin_obesity is covering for cov_num_bmi & cov_cat_bmi_groups,
+  ## cov_cat_hba1c_mmol_mol is covering for cov_num_hba1c_mmol_mol
+  ## cov_cat_tc_hdl_ratio is covering for cov_num_tc_hdl_ratio
+  ## spline(cov_num_age) is covering for cov_cat_age
+  setdiff(c("cov_num_bmi", "cov_cat_bmi_groups", "cov_num_hba1c_mmol_mol", "cov_cat_age", "cov_num_tc_hdl_ratio"
+            )) 
+# print(covariate_names)
+
+# Construct model formula dynamically
+ps_formula <- as.formula(paste("exp_bin_treat ~ rcs(cov_num_age, age_knots) +", paste(covariate_names, collapse = " + "), "+ strata(strat_cat_region)"))
```

Comment:
`strata` returns an error for me in R but seems to be going through the OpenSAFELY command line tool fine. The command in `rms` is called `strat`:

```suggestion
ps_formula <- as.formula(paste("exp_bin_treat ~ rcs(cov_num_age, age_knots) +", paste(covariate_names, collapse = " + "), "+ strat(strat_cat_region)"))
```

Reply:
Thanks for spotting this - changed.